### DISCUSSION TITLE
What is the best way to determine if a download has completed?

### DISCUSSION THREAD

**Discussion by kslr** [anacrolix/torrent#895] [Q&A] [answered]
hi all. I need to generate thumbnails as soon as the download completes, so I need to make sure the file has been written. I see that there are different ways to do this:

https://github.com/anacrolix/torrent/issues/168#issuecomment-323483820
https://github.com/anacrolix/torrent/issues/828#issuecomment-1521677133

```go
1. BytesCompleted == TotalLength
2. BytesMissing == 0
```

Which one is more appropriate for this scenario?
Thanks to all of you

**Comment by anacrolix** [comment, marked as answer]
My strategy for thumbnails is to provide the file as an input over HTTP, and let it use the streaming support in anacrolix/torrent to just fetch what it needs. It typically results in thumbnails appearing long before file is completely downloaded.

Here's the implementation I use: https://github.com/anacrolix/webtorrent-public/blob/master/services/poster.go.

See how it's done in https://www.coveapp.info/ for a demo.
